spacy/vocab.pyx

```diff
@@ -357,6 +365,8 @@ cdef class Vocab:
 
         DOCS: https://spacy.io/api/vocab#prune_vectors
         """
+        if not isinstance(self.vectors, Vectors):
+            raise ValueError(Errors.E849.format(action="prune", vectors_type=type(self.vectors)))
         if self.vectors.mode != VectorsMode.default:
```

Comment:
These three methods that now raise `E849`, could we have them somehow as methods on `(Base)Vectors` directly? So that other implementations have the option to support these operations, and we don't have this awkward type-checking done by the vocab?

Reply:
I agree that it's clunky, but the implementations in `Vocab` are so tied to the specific `Vectors` implementation that I think this would be difficult to do. (If the `Vocab` versions work with your custom vectors, you may as well be using `Vectors` instead.)

I think if you need pruning, deduplication, etc. for your custom vectors, you could implement this for your vocab/pipeline outside of `Vocab`? I don't think there's any requirement that this needs to be implemented as a `Vocab` method, you'd just need to pass in the full `Vocab` in order to have access to the lexical info, etc. if you need it.

Reply:
It's the fact that these implementations in `Vocab` are so tied to the specific `Vectors` implementation that made me wonder whether we shouldn't consider moving those implementations to `Vectors` instead. Then like you say you can just pass the `vocab` to that method and let the (custom) vectors do everything else?

Reply:
In general I think they could be separate `util`-style methods instead that take a `Vocab` arg, but I don't want to change the current `Vocab` API (at least not in v3). I don't think it's a huge deal to leave it like it is, since for the most part we're the only ones using these methods internally, and basically only for `init vectors`.